jina/drivers/evaluate.py

```diff
@@ -0,0 +1,58 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from typing import Iterable
+from .helper import DocGroundtruthPair
+from . import BaseExecutableDriver
+
+
+class BaseEvaluationDriver(BaseExecutableDriver):
+    def __init__(self, executor: str = None,
+                 method: str = 'evaluate',
+                 *args,
+                 **kwargs):
+        super().__init__(executor, method, *args, **kwargs)
+
+    def __call__(self, *args, **kwargs):
+        assert len(self.req.docs) == len(self.req.groundtruths)
+        docs_groundtruths = [DocGroundtruthPair(doc, groundtruth) for doc, groundtruth in
+                             zip(self.req.docs, self.req.groundtruths)]
+        self._traverse_apply(docs_groundtruths, *args, **kwargs)
+
+    def _apply_all(self, docs_groundtruths: Iterable[DocGroundtruthPair],
+                   context_doc_groundtruth: DocGroundtruthPair,
+                   *args,
+                   **kwargs) -> None:
+        pass
+
+
+class RankingEvaluationDriver(BaseEvaluationDriver):
+    """Drivers inherited from this Driver will bind :meth:`evaluate` by default
+    """
+
+    def __init__(self,
+                 id_tag='id',
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.id_tag = id_tag
+
+    @property
+    def id(self):
+        if self.pea:
+            return self.pea.name
+        else:
+            return self.__class__.__name__
+
+    def _apply_all(self,
+                   docs_groundtruths: Iterable[DocGroundtruthPair],
```

Comment:
Since it overwrites an super function, if should still be named `docs`, not? Imagine the iterator once will call `self._apply_all(docs=docs)` then this function will break.

Reply:
Good point, or is thee also an alias?

Reply:
Alias? You mean an alias decorator? I have never seen one.

Reply:
Similar to deprecated?

Reply:
Sure, you could do that. Or you could just do
```
def _apply_all(self, docs: Iterable...):
    self.check_ground_truth(docs)

def check_ground_truth(self, docs_groundtruths: Iterable...):
   assert ...
```
Which would be more verbose and better readable, since it makes the mapping from `docs` to `docs_groundtruths` very explicit.